hpx::parallel::fill fails compiling

@stevenrbrandt reported:

> I have this code that used to compile. Essentially, it's just a call to parallel fill. Now it gives an error. Is it a bug? Thanks.
> 
> The problem line of code is this:
hpx::parallel::fill(hpx::parallel::execution::par.on(exec),vd.begin(),vd.end(),2.0f);
>
> Here's the compilation with error messages...
```
/usr/bin/c++ fx.cpp `pkg-config --cflags hpx_application_release` -c In file included from /usr/local/include/hpx/parallel/algorithm.hpp:22,
                  from /usr/local/include/hpx/algorithm.hpp:10,
                  from /usr/local/include/hpx/hpx.hpp:10,
                  from fx.cpp:1:
/usr/local/include/hpx/parallel/algorithms/fill.hpp: In instantiation of ‘typename hpx::parallel::util::detail::algorithm_result<ExPolicy, 
void>::type hpx::parallel::v1::fill(ExPolicy&&, FwdIter, FwdIter, const
T&) [with ExPolicy =
hpx::execution::parallel_policy_shim<hpx::compute::host::block_executor<>,
hpx::execution::static_chunk_size>; FwdIter = __gnu_cxx::__normal_iterator<float*, std::vector<float> >; T = float; int _concept_requires_217 = 42; typename
std::enable_if<((_concept_requires_637 == 43) || (hpx::is_execution_policy<T>::value && hpx::traits::is_forward_iterator<FwdIter>::value)), int>::type
_concept_check_217 = 0; typename
hpx::parallel::util::detail::algorithm_result<ExPolicy, void>::type =
void]’:
fx.cpp:38:88:   required from here
/usr/local/include/hpx/parallel/algorithms/fill.hpp:232:29: error: 
return-statement with a value, in function returning ‘hpx::parallel::util::detail::algorithm_result_impl<hpx::execution::parallel_policy_shim<hpx::compute::host::block_executor<>,
hpx::execution::static_chunk_size>, void>::type’ {aka ‘void’} [-fpermissive]
   232 |         return detail::fill_(
       |                ~~~~~~~~~~~~~^
   233 |             std::forward<ExPolicy>(policy), first, last, value, is_segmented());
       | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
And here's the code...
```
#include <hpx/hpx.hpp>
#include <hpx/include/parallel_fill.hpp> 
#include <hpx/include/compute.hpp> 
#include <hpx/include/parallel_executors.hpp>

#include <fstream>
#include <sstream>

void prvec(std::vector<float> v){
     for(auto i : v)
       std::cout << i << " ";
     std::cout << std::endl;

};

hpx::compute::host::block_executor<> getexec(){
     auto host_targets = hpx::compute::host::get_local_targets();
     typedef hpx::compute::host::block_executor<> executor_type;
     executor_type exec(host_targets);

     for(auto host : host_targets)
         std::cout << host.get_locality() << std::endl;

     return exec;
}

void fill_example(){
     auto exec = getexec();

     std::vector<float> vd;
     for(int i=0;i<5;i++) vd.push_back(1.f);
     prvec(vd);
hpx::parallel::fill(hpx::parallel::execution::par.on(exec),vd.begin(),vd.end(),2.0f);
     prvec(vd);
}
```

--- a/libs/parallelism/algorithms/include/hpx/parallel/algorithms/fill.hpp
+++ b/libs/parallelism/algorithms/include/hpx/parallel/algorithms/fill.hpp
@@ -221,7 +221,7 @@ namespace hpx { namespace parallel { inline namespace v1 {
     // clang-format on
     HPX_DEPRECATED_V(
         1, 6, "hpx::parallel::fill is deprecated, use hpx::fill instead")
-        typename util::detail::algorithm_result<ExPolicy>::type
+        typename util::detail::algorithm_result<ExPolicy, FwdIter>::type
         fill(ExPolicy&& policy, FwdIter first, FwdIter last, T const& value)
     {
         static_assert((hpx::traits::is_forward_iterator<FwdIter>::value),

--- a/libs/parallelism/algorithms/tests/regressions/CMakeLists.txt
+++ b/libs/parallelism/algorithms/tests/regressions/CMakeLists.txt
@@ -6,6 +6,7 @@
 
 set(tests
     count_3646
+    fill_executor_5016
     for_each_annotated_function
     for_loop_2281
     minimal_findend

--- a/libs/parallelism/algorithms/tests/regressions/fill_executor_5016.cpp
+++ b/libs/parallelism/algorithms/tests/regressions/fill_executor_5016.cpp
@@ -1,0 +1,44 @@
+//  Copyright (c) 2020 Steven R. Brandt
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// #5016: hpx::parallel::fill fails compiling
+
+// suppress deprecation warnings for algorithms
+#define HPX_HAVE_DEPRECATION_WARNINGS_V1_6 0
+
+#include <hpx/hpx_main.hpp>
+
+#include <hpx/include/compute.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/parallel_fill.hpp>
+#include <hpx/modules/testing.hpp>
+
+#include <algorithm>
+#include <vector>
+
+void fill_example()
+{
+    hpx::execution::parallel_executor exec;
+
+    std::vector<float> vd(5);
+    hpx::parallel::fill(
+        hpx::execution::par.on(exec), vd.begin(), vd.end(), 2.0f);
+
+    std::vector<float> vd1(5);
+    hpx::fill(hpx::execution::par.on(exec), vd1.begin(), vd1.end(), 2.0f);
+
+    std::vector<float> expected(5);
+    std::fill(expected.begin(), expected.end(), 2.0f);
+
+    HPX_TEST(vd == expected);
+    HPX_TEST(vd1 == expected);
+}
+
+int main()
+{
+    fill_example();
+    return hpx::util::report_errors();
+}

